src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java

```diff
@@ -221,6 +222,52 @@ public void setTotalExperience(int exp) {
         this.experienceTotal = exp;
     }
 
+    /**
+     * {@link EntityPlayer#addExperienceLevel(int)} and
+     * {@link EntityPlayer#removeExperienceLevel(int)} don't update the total
+     * experience. This recalculates it for plugins to properly make use of it.
+     *
+     * @see #onAddExperienceLevels(int, CallbackInfo)
+     */
+    private void recalculateTotalExperience() {
+        if (!this.dontRecalculateExperience) {
+            this.experienceTotal = this.xpAtLevel(this.experienceLevel) + Math.round(this.experience * this.xpBarCap());
```

Comment:
Apparently this calculation gets pretty wonky... I'll start by addressing the third point on recalculating `this.experience`.

### Should `this.experience` be recalculated?

Let's say you were on a level that requires 8 XP to move on to the next level, and you had 6 XP towards it. If you change levels using `/xp`, Minecraft preserves the percentage of XP in your XP bar. So if someone pulled you down to a level that required 4 XP to move on, you would have 3 XP on your way to the next level.

On the other hand, if they first pulled you down to a level that needed 6 XP to move on (in a separate command), you would have 75% of 6 XP. Vanilla Minecraft doesn't care about it; when they pull you down to the level needing 4 XP, you're in exactly the same state as above. On the other hand, when tweaking `this.experience` in the recalculation of `totalExperience`, you would either end up with 2 of 4 XP (if you round down; 75% would go to 4/6, or 67%, and then that would go down to 2/4, or 50%) or 3 of 4 XP (if you round to the nearest whole number; 75% would go to 5/6, or 83%, and then that would back to 75%) - though the latter wouldn't necessarily work out either if there were more steps.

In this case, I left `this.experience` alone because I see `totalExperience` as based off `experience` and `experienceLevel`; if the experience changes, `totalExperience` will properly follow it.

### Round to the nearest value or down?
Vanilla Minecraft would say round down. If you had 5 XP on a level that needed 6 XP to continue, you would be 83% of the way to the next level. If you then went down to a level that needed 2 XP to continue, you would still be 83% of the way - except the next experience point you collect would progress you 33% of the way, pushing you slightly over. If you round to the nearest whole number, you would be pushed on to the next level, which is inconsistent with vanilla.

The only problem? Floating point inaccuracies. In my testing when I had rounding down, I gave myself 100 XP and then ran `/xp 0L <player name>` to simply call `recalculateTotalExperience()`. Then the `totalExperience` was recalculated, but was off slightly due to the floating point error: I ended up with 99 XP. I had supposedly been 9 XP through the level, and it needed 21 XP to continue, but the code actually said I was `8.999997` XP through the level. Rounding down then stole an XP point.

Another option is a threshold. That would work out nicely, except that I've found Minecraft's `addExperience()` method exaggerates the floating point inaccuracies to the point where a single XP point can be smaller on the scale than its inaccuracy.

All in all, I have no idea what would work best, considering that we still need Sponge to work with existing save files that used the inaccurate floating point values (otherwise I could completely rip out the current XP calculation system and replace it with something more accurate). Any ideas would be greatly appreciated.